src/dialog-editor/components/box/boxComponent.ts

```diff
@@ -0,0 +1,140 @@
+import * as ng from 'angular';
+import * as _ from 'lodash';
+
+/**
+ * Controller for the Dialog Editor box component
+ * @memberof miqStaticAssets
+ * @ngdoc controller
+ * @name BoxController
+ */
+class BoxController {
+  public sortableOptionsBox: any;
+  public sortableOptionsFields: any;
+  public service: any;
+  public dialogTabs: any;
+
+  /*@ngInject*/
+  constructor(private DialogEditor: any, private DialogEditorModal: any) {
+  }
+
+  /**
+   * Load service to be able to access it form the template.
+   * Load status of tabs.
+   * @memberof BoxController
+   * @function $onInit
+   */
+  public $onInit() {
+    this.service = this.DialogEditor;
+    this.dialogTabs = this.DialogEditor.getDialogTabs();
+    // Rules for Drag&Drop sorting of boxes
+    this.sortableOptionsBox = {
+      axis: 'y',
+      cancel: '.nosort',
+      cursor: 'move',
+      opacity: 0.5,
+      revert: 50,
+      stop: (e: any, ui: any) => {
+        let sortedBox = ui.item.scope().$parent.tab.dialog_groups;
+        // update indexes of other boxes after changing their order
+        this.DialogEditor.updatePositions(sortedBox);
+      },
+    };
+    // Rules for Drag&Drop sorting of elements inside of boxes
+    this.sortableOptionsFields = {
+      axis: 'y',
+      cancel: '.nosort',
+      cursor: 'move',
+      revert: 50,
+      stop: (e: any, ui: any) => {
+        let sortedField = ui.item.scope().$parent.box.dialog_fields;
+        // update indexes of other fields after changing their order
+        this.DialogEditor.updatePositions(sortedField);
+      },
+    };
+  }
+
+  /**
+   * Add a new box to the list.
+   * The new box is automatically appended to the last position of the list
+   * @memberof BoxController
+   * @function addBox
+   */
+  public addBox() {
+    this.dialogTabs[this.DialogEditor.activeTab].dialog_groups
+      .push(
+        {
+          description: __('Description'),
+          label: __('Label'),
+          display: 'edit',
+          position: 0,
+          dialog_fields: [],
+        }
+      );
+    // update indexes of other boxes after adding a new one
+    this.DialogEditor.updatePositions(
+      this.dialogTabs[this.DialogEditor.activeTab].dialog_groups
+    );
+  }
+
+  /**
+   * Remove box and all its content from the dialog.
+   * @memberof BoxController
+   * @function removeBox
+   * @param {number} id as index of removed box
+   */
+  public removeBox(id: number) {
+    _.remove(
+      this.dialogTabs[this.DialogEditor.activeTab].dialog_groups,
+      function(box) {
```

Comment:
Thanks, I've used the arrow function so far on places, where I had to pass `this` (http://stackoverflow.com/questions/16157839/typescript-this-inside-a-class-method helped). 

I'll replace also the other functions, you've mentioned 👍